cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+      = dolfinx::MPI::local_range(rank, num_vertices_global, size);
+
+  std::array<std::uint64_t, 2> local_range{(std::uint64_t)_local_range[0],
+                                           (std::uint64_t)_local_range[1]};
+  std::uint64_t num_vertices_local = local_range[1] - local_range[0];
+
+  std::array<std::int64_t, 2> _cell_range
+      = dolfinx::MPI::local_range(rank, num_cells_global, size);
+
+  std::array<std::uint64_t, 2> cell_range{(std::uint64_t)_cell_range[0],
+                                          (std::uint64_t)_cell_range[1]};
+  std::uint64_t num_cells_local = cell_range[1] - cell_range[0];
+
+  std::vector<int64_t> input_global_indices(num_vertices_local);
+  std::vector<T> x(num_vertices_local * 3);
+  std::vector<int64_t> array(num_cells_local * num_dofs_per_cell);
+  std::vector<int32_t> offsets(num_cells_local + 1);
+
+  {
+    adios2::Variable<std::int64_t> var_input_global_indices
+        = io->InquireVariable<std::int64_t>("input_global_indices");
+
+    adios2::Variable<T> var_x = io->InquireVariable<T>("x");
+
+    adios2::Variable<std::int64_t> var_topology_array
+        = io->InquireVariable<std::int64_t>("topology_array");
+
+    adios2::Variable<std::int32_t> var_topology_offsets
+        = io->InquireVariable<std::int32_t>("topology_offsets");
+
+    if (var_input_global_indices)
+    {
+      var_input_global_indices.SetSelection(
+          {{local_range[0]}, {num_vertices_local}});
+      engine->Get(var_input_global_indices, input_global_indices.data(),
+                  adios2::Mode::Deferred);
+    }
+
+    if (var_x)
+    {
+      var_x.SetSelection({{local_range[0], 0}, {num_vertices_local, 3}});
+      engine->Get(var_x, x.data(), adios2::Mode::Deferred);
+    }
+
+    if (var_topology_array)
+    {
+      var_topology_array.SetSelection({{cell_range[0] * num_dofs_per_cell},
+                                       {cell_range[1] * num_dofs_per_cell}});
+      engine->Get(var_topology_array, array.data(), adios2::Mode::Deferred);
+    }
+
+    if (var_topology_offsets)
+    {
+      var_topology_offsets.SetSelection({{cell_range[0]}, {cell_range[1] + 1}});
+      engine->Get(var_topology_offsets, offsets.data(), adios2::Mode::Deferred);
+    }
+
+    engine->EndStep();
+
+    std::int32_t cell_offset = offsets[0];
+    for (auto offset = offsets.begin(); offset != offsets.end(); ++offset)
+      *offset -= cell_offset;
+  }
+
+  std::vector<T> x_reduced(num_vertices_local * dim);
+  for (std::uint32_t i = 0; i < num_vertices_local; ++i)
+  {
+    for (std::uint32_t j = 0; j < (std::uint32_t)dim; ++j)
+      x_reduced[i * dim + j] = x[i * 3 + j];
+  }
+
+  fem::CoordinateElement<T> element
+      = fem::CoordinateElement<T>(cell_type, degree, lagrange_variant);
+
+  std::array<std::size_t, 2> xshape = {num_vertices_local, (std::uint32_t)dim};
+  auto part = mesh::create_cell_partitioner(mesh::GhostMode::shared_facet);
+
+  mesh::Mesh<T> mesh = mesh::create_mesh(comm, comm, array, element, comm,
```

Comment:
As this will be a low level function for constructing arrays of data it should just return the low level arrays, the higher level read_mesh can do the actual creation of the mesh.